### PR TITLE
Adding changes to collect OVN audit logs LOG-1526

### PR DESCRIPTION
**Pull request by ajaygupta978**
### Description
This PR address collection of OVN audit logs by fluentd. 

/cc @vimalk78 
/assign @jcantrill 

### Links

- Depending on PR(s): 
- Bugzilla:
- Github issue:
- JIRA: https://issues.redhat.com/browse/LOG-1526
- Enhancement proposal:

### CODE REVIEW

**Review by vimalk78 — CHANGES_REQUESTED** [review]
please 
add unit tests for the new changes
add functional test for the OVN audit logs
update existing unit tests. does the OVN audit logs need to be captured everytime ? can logging operator detect presence/absence of OVN before generating config for capturing OVN audit logs?

**Comment by ajaygupta978** [comment]
Input OVN Audit log sample

`2021-07-06T08:26:58.687Z|00004|acl_log(ovn_pinctrl0)|INFO|name="verify-audit-logging_deny-all", verdict=drop, severity=alert:icmp,vlan_tci=0x0000,dl_src=0a:58:0a:81:02:12,dl_dst=0a:58:0a:81:02:14,nw_src=10.129.2.18,nw_dst=10.129.2.20,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0`

OVN logs in Elastic search

```{
        "_index": "audit-000001",
        "_type": "_doc",
        "_id": "NTkzYTBiMDEtZDc5Mi00ZTQwLWI5OWUtMGMyNThjOWUyMDgz",
        "_score": 1.0185696,
        "_source": {
          "message": "2021-07-06T08:26:58.687Z|00004|acl_log(ovn_pinctrl0)|ERROR|name=\"verify-audit-logging_deny-all\", verdict=drop, severity=alert: icmp,vlan_tci=0x0000,dl_src=0a:58:0a:81:02:12,dl_dst=0a:58:0a:81:02:14,nw_src=10.129.2.18,nw_dst=10.129.2.20,nw_tos=0,nw_ecn=0,nw_ttl=64,icmp_type=8,icmp_code=0",
          "@timestamp": "2021-07-06T08:26:58.687Z",
          "level": "info",
          "hostname": "ip-10-0-161-225.us-west-1.compute.internal",
          "pipeline_metadata": {
            "collector": {
              "ipaddr4": "10.0.161.225",
              "inputname": "fluent-plugin-systemd",
              "name": "fluentd",
              "received_at": "2021-07-15T08:26:45.077691+00:00",
              "version": "1.7.4 1.6.0"
            }
          },
          "viaq_msg_id": "NTkzYTBiMDEtZDc5Mi00ZTQwLWI5OWUtMGMyNThjOWUyMDgz",
          "kubernetes": {
            
          }
        }
      }

File: test/functional/framework.go [hunk 1]
```diff
@@ -389,6 +399,10 @@ func (f *FluentdFunctionalFramework) Readk8sAuditLogsFrom(outputName string) ([]
 	return f.ReadLogsFrom(outputName, k8sAuditLog)
 }
 
+func (f *FluentdFunctionalFramework) ReadovnAuditLogsFrom(outputName string) ([]string, error) {
```
Comment (vimalk78) [inline test/functional/framework.go:402]: change method name to `ReadOvnAuditLogsFrom` , change `ovn` to `Ovn`

**Review by vimalk78 — COMMENTED** [review]
As mentioned in earlier comment, we can optimize and not capture ovn logs if ovn logs will not be present every time on the node.
i.e. is OVN present on cluster under a condition, or is it always present out of the box? 
If it is not present, we could check the same condition in operator.

File: test/functional/framework.go [hunk 2]
```diff
@@ -354,6 +358,12 @@ func (f *FluentdFunctionalFramework) WriteMessagesToOpenshiftAuditLog(msg string
 	return f.WriteMessagesToLog(msg, numOfLogs, filename)
 }
 
+func (f *FluentdFunctionalFramework) WriteMessagesToOVNAuditLog(msg string, numOfLogs int) error {
+
+	filename := fmt.Sprintf("%s/acl-audit-log.log", fluentdLogPath[ovnAuditLog])
```
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply test/functional/framework.go:363]: @vimal this is the actual ovn  audit log file name in production system so I used same name in test cases.

File: test/functional/framework.go (hunk 1 above)
Comment (ajaygupta978) [reply test/functional/framework.go]: done

File: test/helpers/types/types.go [hunk 3]
```diff
@@ -178,6 +179,18 @@ type AuditLinux struct {
 	RecordID string `json:"record_id,omitempty"`
 }
 
+// OVN Audit log
```
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply test/helpers/types/types.go:182]: @vimalk78 done. Please review it again.

File: pkg/generators/forwarding/fluentd/fluent_conf_test.go [hunk 4]
```diff
@@ -2837,8 +2885,16 @@ var _ = Describe("Generating fluentd config", func() {
 						key PRIORITY
 						pattern ^7$
 					</exclude>
-				</filter>
-				<match journal>
+          </filter>
```
Comment (vimalk78) [inline pkg/generators/forwarding/fluentd/fluent_conf_test.go:2888]: please fix the indentation

File: pkg/generators/forwarding/fluentd/fluent_conf_test.go [hunk 5]
```diff
@@ -2837,8 +2885,16 @@ var _ = Describe("Generating fluentd config", func() {
...
+
+          <filter ovn-audit.log**>
+          @type record_modifier
```
Comment (vimalk78) [inline pkg/generators/forwarding/fluentd/fluent_conf_test.go:2891]: fix the indentation

File: pkg/generators/forwarding/fluentd/templates.go [hunk 6]
```diff
@@ -141,13 +142,21 @@ const fluentConfTemplate = `{{- define "fluentConf" -}}
 
   ## filters
   <filter journal>
-    @type grep
+	@type grep
```
Comment (vimalk78) [inline pkg/generators/forwarding/fluentd/templates.go:145]: please fix indentation

File: pkg/generators/forwarding/fluentd/templates.go [hunk 7]
```diff
@@ -141,13 +142,21 @@ const fluentConfTemplate = `{{- define "fluentConf" -}}
...
     <exclude>
       key PRIORITY
       pattern ^7$
     </exclude>
   </filter>
 
+  <filter ovn-audit.log**>
+  @type record_modifier
```
Comment (vimalk78) [inline pkg/generators/forwarding/fluentd/templates.go:153]: please fix indentation

**Review by vimalk78 — COMMENTED** [review]
please fix indentation, in unit tests also

File: pkg/generators/forwarding/fluentd/templates.go [hunk 8]
```diff
@@ -517,6 +526,25 @@ const inputSourceOpenShiftAuditTemplate = `{{- define "inputSourceOpenShiftAudit
...
+  @type tail
+  @id ovn-audit-input
+  @label @MEASURE
+  path "/var/log/ovn/acl-audit-log.log"
+  pos_file "/var/lib/fluentd/pos/acl-audit-log.pos"
+  tag ovn-audit.log
+  refresh_interval 5
+  rotate_wait 5
+  read_from_head true
+  <parse>
```
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply pkg/generators/forwarding/fluentd/templates.go:541]: @jcantrill I am not sure If I understand but apart from these tests I also tested it manually by creating a log file on a node which is read by fluentd. I didn't notice default message field is added. Fluentd gets original log line without `message` field. Do you still want me to test something else?

File: pkg/k8shandler/fluentd.go [hunk 9]
```diff
@@ -417,6 +417,7 @@ func newFluentdPodSpec(cluster *logging.ClusterLogging, trustedCABundleCM *v1.Co
 		{Name: "varlogpods", ReadOnly: true, MountPath: "/var/log/pods"},
 		{Name: "varlogjournal", ReadOnly: true, MountPath: "/var/log/journal"},
 		{Name: "varlogaudit", ReadOnly: true, MountPath: "/var/log/audit"},
+		{Name: "varlogovn", ReadOnly: true, MountPath: "/var/log/ovn"},
```
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply pkg/k8shandler/fluentd.go]: ok done.

File: pkg/k8shandler/fluentd.go [hunk 10]
```diff
@@ -506,6 +507,7 @@ func newFluentdPodSpec(cluster *logging.ClusterLogging, trustedCABundleCM *v1.Co
 			{Name: "varlogpods", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/var/log/pods"}}},
 			{Name: "varlogjournal", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/var/log/journal"}}},
 			{Name: "varlogaudit", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/var/log/audit"}}},
+			{Name: "varlogovn", VolumeSource: v1.VolumeSource{HostPath: &v1.HostPathVolumeSource{Path: "/var/log/ovn"}}},
```
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply pkg/k8shandler/fluentd.go]: done

File: pkg/generators/forwarding/fluentd/output_conf_legacy_test.go [hunk 11]
```diff
@@ -637,6 +660,22 @@ var _ = Describe("Generating fluentd legacy output store config blocks", func()
...
+				  path "/var/log/ovn/acl-audit-log.log"
+				  pos_file "/var/lib/fluentd/pos/acl-audit-log.pos"
+				  tag ovn-audit.log
+				  refresh_interval 5
+				  rotate_wait 5
+				  read_from_head true
+				  <parse>
+					@type regexp
+					expression /^(?<message>.*)$/
+				  </parse>
```
Comment (vimalk78) [inline pkg/generators/forwarding/fluentd/output_conf_legacy_test.go:677]: indentation is still broken

**Review by vimalk78 — COMMENTED** [review]
fix indentation

File: pkg/generators/forwarding/fluentd/templates.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ajaygupta978) [reply pkg/generators/forwarding/fluentd/templates.go:541]: @jcantrill Thanks for suggestion. I  tested with `none` and it works as expected so we don't need regex.

File: pkg/k8shandler/fluentd.go [hunk 12]
```diff
@@ -37,6 +37,44 @@ const (
 	fluentdRequiredESVersion = "6"
 	logVolumeMountName       = "varlog"
 	logVolumePath            = "/var/log"
+	logContainers            = "varlogcontainers"
```
Comment (jcantrill) [inline pkg/k8shandler/fluentd.go:40]: Thanks for this change but was beyond my ask.  Its an improvement none the less

**Pull request merged**